File: IOSViperSample/IOSViperSample/chapters/ChaptersRepositoryImpl.swift
//
// Created by Ivan Phytsyk on 9/29/18.
// Copyright (c) 2018 Ivan Phytsyk. All rights reserved.
//

import Foundation

class ChaptersRepositoryImpl: ChapterRepository {
    
    let chapters: [Chapter] = [Chapter(title: "About Book", description: "Short description about author and this book content", type: ChapterType.about_book)
        , Chapter(title: "Classes", description: "Explains about classes, there differences in both languages and unique features", type: ChapterType.classes_interfaces)
        , Chapter(title: "Layouts", description: "ConstraintLayout and AutoLayout as an approach of interface building", type: ChapterType.layouts)
        , Chapter(title: "Functions", description: "Functions syntax, usage. Lambdas and high-level functions.", type: ChapterType.functions)
        , Chapter(title: "Collections", description: "Short brief about collections types and their usage", type: ChapterType.collections)
        , Chapter(title: "Data types", description: "Basic data types, optionals and nullables", type: ChapterType.data_types)
        , Chapter(title: "Asynchronous execution", description: "Starting tasks in background and sending result back to UI thread", type: ChapterType.asynchronous)
        , Chapter(title: "Project structure", description: "Files and folders projects consist of", type: ChapterType.project_structure)
        , Chapter(title: "View Controllers", description: "Activity and UIViewController purpose, layout and lifecycles", type: ChapterType.viewcontroller)
        , Chapter(title: "Dimensions", description: "Widgets sizes units: point and dp", type: ChapterType.dimensions)
        , Chapter(title: "Layout", description: "ConstraintLayout and AutoLayout. Creating layout with constraints to support various screen sizes", type: ChapterType.layouts)
        , Chapter(title: "Lists", description: "Recycled widget list. Delegators and Adapters.", type: ChapterType.lists)
        , Chapter(title: "Do we need the same architecture", description: "My opinion about mobile architecture", type: ChapterType.why)
        , Chapter(title: "V.I.P.E.R", description: "My architecture approach describing", type: ChapterType.viper)
        , Chapter(title: "Testing", description: "Unit tests for business logic.", type: ChapterType.testing)
        
    ]
    
    func getChapter(_ title: String) -> Chapter {
        return chapters.first(where: {
            $0.title == title
        })!
    }
    
    func getChapters() -> Array<Chapter> {
        return chapters
    }
    
    
}

File: AndroidViperSample/app/src/main/java/fytsyk/com/vipersample/chapters/ChaptersRepositoryImpl.kt
package fytsyk.com.vipersample.chapters

import fytsyk.com.vipersample.chapters.viper.Chapter
import fytsyk.com.vipersample.chapters.viper.ChapterRepository
import fytsyk.com.vipersample.chapters.viper.ChapterType

/**
 * Just hardcoded list of business objects. Usually it uses network/db for data fetching or saving
 */
class ChaptersRepositoryImpl : ChapterRepository {

    val chapterList = listOf(Chapter("About Book", "Short    about author and this book content", ChapterType.about_book)
            , Chapter("Classes", "Explains about classes, there differences in both languages and unique features", ChapterType.classes_interfaces)
            , Chapter("Layouts", "ConstraintLayout and AutoLayout as an approach of interface building", ChapterType.layouts)
            , Chapter("Functions", "Functions syntax, usage. Lambdas and high-level functions.", ChapterType.functions)
            , Chapter("Collections", "Short brief about collections types and their usage", ChapterType.collections)
            , Chapter("Data types", "Basic data types, optionals and nullables", ChapterType.data_types)
            , Chapter("Asynchronous execution", "Starting tasks in background and sending result back to UI thread", ChapterType.asynchronous)
            , Chapter("Project structure", "Files and folders projects consist of", ChapterType.project_structure)
            , Chapter("View Controllers", "Activity and UIViewController purpose, layout and lifecycles", ChapterType.viewcontroller)
            , Chapter("Dimensions", "Widgets sizes units  point and dp", ChapterType.dimensions)
            , Chapter("Layout", "ConstraintLayout and AutoLayout. Creating layout with constraints to support various screen sizes", ChapterType.layouts)
            , Chapter("Lists", "Recycled widget list. Delegators and Adapters.", ChapterType.lists)
            , Chapter("Do we need the same architecture", "My opinion about mobile architecture", ChapterType.why)
            , Chapter("V.I.P.E.R", "My architecture approach describing", ChapterType.viper)
            , Chapter("Testing", "Unit tests for business logic.", ChapterType.testing))

    override fun getChapters(): List<Chapter> {
        return chapterList
    }

    override fun getChapter(title: String): Chapter {
        return chapterList.find { it.title == title }!!
    }
}